### PR TITLE
allow inclusion of environment variables during deployment

### PR DESCRIPTION
**Pull request by kevinushey**
### TODO

- [x] Consider updating the required version of `rsconnect`; it looks like `envVars` have been supported since 1.0.0.
- [x] Add placeholder UI for case where no environment variables are available, or `.Renviron` does not exist.
- [x] Add a help link into the environment variables modal (https://docs.posit.co/connect/admin/process-management/index.html#environment-variables)
- [x] Consider whether we want to allow auto-updating of environment variables. (I'm leaning towards _no_, and that users should have to manually select and publish environment variables each time, but can be convinced otherwise!)

### Intent

Addresses https://github.com/rstudio/rstudio/issues/13032.

### Approach

Recent releases of `rsconnect` include support for publishing environment variables alongside a particular deployment. This is typically useful for secrets / API keys which the application might need in order to access external web services. This PR adds some basic UI to make this possible. Example:

Screenshot 2024-09-09 at 3 31 30 PM

Note the new `Add environment variables...` button. Clicking that brings up a simple modal dialog:

Screenshot 2024-09-09 at 3 31 57 PM

The names of environment variables are pulled from either the project's `.Renviron`, or from the user `.Renviron`. Currently, names are filtered to those which "look like" secret keys; we might reconsider this design in the future if needed.

After selecting some environment variables and clicking Ok, we add a small blurb to indicate these variables were set:

Screenshot 2024-09-09 at 3 33 09 PM

These environment variable names then become part of the `deployApp()` call generates to start the deployment.

The empty placeholder is fairly minimal; the hope is that interested users will click and appropriate help link. We can add more information if needed.

Screenshot 2024-09-10 at 11 47 58 AM

### Automated Tests

N/A

### QA Notes

Test via notes in https://github.com/rstudio/rstudio/issues/13032.

### Documentation

To be added.

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/rserver-dev.in [hunk 1]
```diff
@@ -35,5 +35,5 @@ trap cleanup SIGINT
 RS_CRASH_HANDLER_PATH="$(pwd)/server/crash-handler-proxy/crash-handler-proxy" \
    RS_CRASHPAD_HANDLER_PATH="${RSTUDIO_TOOLS_ROOT}/crashpad/crashpad/out/Default/crashpad_handler" \
    RS_DB_MIGRATIONS_PATH="${CMAKE_CURRENT_SOURCE_DIR}/server/db" \
-   server/rserver --server-user $(whoami) --config-file conf/rserver-dev.conf "$@"
+   server/rserver --server-user $(whoami) --auth-none=1 --config-file conf/rserver-dev.conf "$@"
```
Comment (kevinushey) [inline src/cpp/rserver-dev.in:38]: I feel like we can skip the authentication step when working in open source development builds now.

File: src/cpp/r/session/RConsoleActions.cpp [hunk 2]
```diff
@@ -200,6 +200,10 @@ Error ConsoleActions::loadFromFile(const FilePath& filePath)
       if (error && !isFileNotFoundError(error))
          return error;
 
+      // bail if file is empty
+      if (contents.empty())
+         return Success();
```
Comment (kevinushey) [inline src/cpp/r/session/RConsoleActions.cpp:205]: I was seeing periodic errors when restoring console actions; this is a small defensive fix I brought into this PR.

File: src/cpp/rserver-dev.in (hunk 1 above)
Comment (gtritchie) [reply src/cpp/rserver-dev.in:38]: Seems reasonable.

File: src/cpp/session/modules/SessionRSConnect.R [hunk 3]
```diff
@@ -13,6 +13,24 @@
...
+      return(character())
+   
+   # Read environment variable names from the file
+   contents <- readLines(environFile, warn = FALSE)
+   pattern <- "^\\s*([\\w_]+)\\s*="
+   matchedLines <- grep(pattern, contents, perl = TRUE, value = TRUE)
+   envVars <- gsub("\\s*=.*", "", matchedLines)
+   
+   # Keep only those tokens which would likely have secrets of interest
+   pattern <- "(?:^|[\\b_])(?:api|key|pass|pat|secret|token)(?:$|[\\b_])"
```
Comment (gtritchie) [inline src/cpp/session/modules/SessionRSConnect.R:30]: For users who have variables that don't match this pattern (you just know that's going to happen) should we make this configurable?

Perhaps a second regular expression (in a setting, initially empty) that specifies additional things to match? Of course, then we end up helping people debug regular expressions, so instead could be an allowlist where additional variables to match can be added.

Also, the other way around, variables that match this built-in pattern that they don't want to show up as a choice; in that case, either let them override the primary regex or maybe a blocklist.

All of this for future consideration, probably more of a server/workbench thing. I don't think we need to over think this for v1.

File: src/cpp/session/resources/dependencies/r-packages.json [hunk 4]
```diff
@@ -251,7 +251,7 @@
             "source": false
         },
         "rsconnect": {
-            "version": "0.8.24",
+            "version": "1.3.1",
```
Comment (gtritchie) [inline src/cpp/session/resources/dependencies/r-packages.json:254]: Paranoia based on past experience: make sure this package version doesn't depend on R >= 4.0.

File: src/gwt/src/org/rstudio/studio/client/rsconnect/ui/RSConnectDeploy.java [hunk 5]
```diff
@@ -154,6 +162,97 @@ public enum ServerType {
       SHINYAPPS
    }
    
+   public static class EnvironmentVariablesDialog extends ModalDialog<List<String>>
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/rsconnect/ui/RSConnectDeploy.java:165]: Minor nit based on the screenshot: when in the "No variables found" mode, be nice to not still have both OK and Cancel buttons, just one button (Ok or Close?). Not worth reworking things if there isn't a trivial solution.

**Review by gtritchie — APPROVED** [review]
Some minor ponderings but LGTM!

File: src/cpp/session/resources/dependencies/r-packages.json (hunk 4 above)
Comment (kevinushey) [reply src/cpp/session/resources/dependencies/r-packages.json:254]: I indeed did double-check that :-)

File: src/gwt/src/org/rstudio/studio/client/rsconnect/ui/RSConnectDeploy.java (hunk 5 above)
Comment (kevinushey) [reply src/gwt/src/org/rstudio/studio/client/rsconnect/ui/RSConnectDeploy.java:165]: Good idea; I'll see how straightforward that is.

**Pull request merged**